### PR TITLE
Add `useRemoteImage` hook #250

### PR DESCRIPTION
**Pull request by thatStrangeGuyThorchain**
- [x] added useRemote hook
- [x] added tests for useRemote hook

closes https://github.com/thorchain/asgardex-electron/issues/250

### CODE REVIEW

File: src/hooks/useRemoteImage.test.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    Object.defineProperty(global.Image.prototype, 'src', {
+      set(src: string) {
+        if (src === LOAD_FAILURE_SRC) {
+          this.onerror()
+        } else if (src === LOAD_SUCCESS_SRC) {
+          this.onload()
+        }
+      }
+    })
+  })
```
Comment (veado) [inline src/hooks/useRemoteImage.test.ts:26]: As with other mocks or stubs, I would recommend to bring `Image` back to it's original "state" in `afterAll` or so.

Comment (thatStrangeGuyThorchain) [reply src/hooks/useRemoteImage.test.ts:26]: @veado nice catch

File: src/hooks/useRemoteImage.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+export const useRemoteImage = (src: string) => {
+  const [remoteImage, setRemoteImage] = useState<RD.RemoteData<string, string>>(RD.pending)
+
+  useEffect(() => {
+    const image = new Image()
+    image.onload = () => setRemoteImage(RD.success(src))
+    image.onerror = () => setRemoteImage(RD.failure(src))
+    image.src = src
+  }, [src])
```
Comment (veado) [inline src/hooks/useRemoteImage.ts:13]: What do you think about to clean-up everything like this?

```
  useEffect(() => {

    const onLoad = () => setRemoteImage(RD.success(src))
    const onError = () => setRemoteImage(RD.failure(src))

    const image = new Image()   
    image.addEventListener('load', onLoad)
    image.addEventListener('error', onError)
    image.src = src

   // clean-up callbacks
    return () => {
       image.removeEventListener('load', onLoad)
       image.removeEventListener('error', onError)
    }
  }, [src])

File: src/components/uielements/assets/assetIcon/AssetIcon.tsx [hunk 3]
```diff
@@ -63,12 +62,7 @@ const AssetIcon: React.FC<Props> = (props: Props): JSX.Element => {
     )
   }, [asset, size])
 
-  return (
-    <>
-      {!error && renderIcon}
-      {error && renderFallbackIcon}
-    </>
-  )
+  return RD.fold(() => <></>, renderFallbackIcon, renderFallbackIcon, renderIcon)(remoteIconImage)
```
Comment (veado) [inline src/components/uielements/assets/assetIcon/AssetIcon.tsx:65]: I would render a fallback icon on error only and for or `initial` and `loading` states `Styled.IconBG`

File: src/hooks/useRemoteImage.ts (hunk 2 above)
Comment (thatStrangeGuyThorchain) [reply src/hooks/useRemoteImage.ts:13]: there is no closures here and image object will cleaned from memory the same time as Unmount. with `image.onload` and `image.onerror` properties. so i believe it's not necessary to use listeners here

File: src/components/uielements/assets/assetIcon/AssetIcon.tsx (hunk 3 above)
Comment (thatStrangeGuyThorchain) [reply src/components/uielements/assets/assetIcon/AssetIcon.tsx:65]: @veado there is no initial state by default behaviour of `useRemoteImage` (initial value is `pending`). for loading state here is a fallback renderer just 'cause i did not have a loading icon which should be placed. 
anyway i will place here whatever icon you say. just replace a few symbols.
just say exactly which ones ?:)

Comment (veado) [reply src/components/uielements/assets/assetIcon/AssetIcon.tsx:65]: For loading state you can use<IconWrapper><Styled.IconBG /><IconWrapper>, where [Styled.IconBG is just a "circle"](https://github.com/thorchain/asgardex-electron/blob/e6461d10c36cd2f68693931f1db5abd735ed5fd5/src/components/uielements/assets/assetIcon/AssetIcon.style.ts#L30-L37) with a size of an icon.

Comment (veado) [reply src/components/uielements/assets/assetIcon/AssetIcon.tsx:65]: Or we could use[ a custom `Spin` provided by `antd`](https://ant.design/components/spin/#components-spin-demo-custom-indicator) - it's up to you ;)

Peek 2020-07-11 19-15

**Pull request merged**